README.md
# About Stubby

Stubby is an application that acts as a local **DNS Privacy stub resolver** (using DNS-over-TLS). Stubby encrypts DNS queries sent 
from a client machine (desktop or laptop) to a DNS Privacy resolver increasing end user privacy. Stubby is in the early stages 
of development but is suitable for technical/advanced users. A more generally user-friendly version is on the way!

Stubby provides DNS Privacy by:

* Running as a daemon
* Listening on the loopback address to send all outgoing DNS queries received on that address out over TLS
* Using a default configuration which provides Strict Privacy and uses a subset
of the available [DNS Privacy servers](https://dnsprivacy.org/test_servers/)

Stubby is developed by the getdns team.

Stubby can be used as a command line tool requiring the user to manually modify the system DNS resolvers. A development version of a Stubby Manager GUI is also available for Windows, see [Stubby Mananger GUI](https://dnsprivacy.org/wiki/display/DP/Stubby+Manager+GUI).

# Documentation

See [Stubby Homepage](https://dnsprivacy.org/dns_privacy_daemon_-_stubby/) for more details

# Dependencies

Stubby uses [getdns](https://getdnsapi.net/) and requires the 1.5.0 release of getdns or later.

It also requires [yaml](http://pyyaml.org/wiki/LibYAML).

# Installing Using a Package Manager

Check to see if stubby, getdns and yaml are available via a package manager for your system:
https://repology.org/metapackage/stubby/versions
https://repology.org/metapackage/getdns/versions
https://repology.org/metapackage/libyaml/versions

* A [Windows Installer](https://dnsprivacy.org/dns_privacy_daemon_-_stubby/installation/windows_installer_for_stubby/) is now available for Stubby. From version 0.4.0 the msi installer includes a development version of a Windowd GUI.
* A Homebrew package for stubby is now available (`brew install stubby`).
* A [GUI for macOS](https://dnsprivacy.org/dns_privacy_daemon_-_stubby/installation/stubby_gui_for_macos/) is also available for testing

If you need to install getdns from source, see the section [at the end of this document.](#building-getdns-from-source)

# Build Stubby from source

A [Dockerfile](./contrib/docker/Dockerfile) is provided in the contrib directory to build a Docker image from source. See [Docker usage docs](./contrib/docker/README.md).

Get the code:
```
git clone https://github.com/getdnsapi/stubby.git
```

Build and install (NOTE: from release 0.3.0 stubby uses cmake)
```
cd stubby
cmake .
make
````
If you get an error `Could NOT find Getdns (missing: GETDNS_LIBRARIES GETDNS_INCLUDE_DIR)` then you need to set the following 2 variable as arguements to `cmake`:
```
cmake -DGETDNS_LIBRARY=<path to libgetdns> -DGETDNS_INCLUDE_DIR=<path to getdns.h header file> .
```
A quick start guide to [using `cmake` with getdns is here.](https://getdnsapi.net/quick-start/cmake-quick-start/)

# Configure Stubby

_It is recommended to use the default configuration file provided which will use 'Strict' privacy mode and spread the DNS 
queries among several of the current DNS Privacy test servers. Note that this file contains both IPv4 and IPv6 addresses. 
This file is installed on *nix systems as /usr/local/etc/stubby/stubby.yml_

The configuration file format is a YAML like format and the name of the file must have an extension of .yml. Essentially the 
configuration options available are the same as the options that can be set on a getdns `context` - Doxygen documentation for 
which is available [here](https://getdnsapi.net/doxygen/group__getdns__context.html). To aid with creating a custom configuration file, an example is given below.

NOTE: As of the 0.1.3 release of Stubby the YAML format replaces the JSON like format used in earlier versions of the configuration file for getdns/stubby. The YAML format is 
more human readable and supports comments allowing options to be easily enabled and disabled. The JSON format is that which is used internally in getdns
(it is the same as the output returned by `stubby -i`) and is still available by directly specifying a file with the name 'stubby.conf' on the command line using the -C option.

## Create Custom Configuration File

Alternatively the configuration file location can be specified on the command line using the `-C` flag. Changes to the configuration file require a restart of Stubby.

The config file below will configure Stubby in the following ways:

*  `resolution_type`: Work in stub mode only (not recursive mode) - required for Stubby operation.
*  `dns_transport_list`: Use TLS only as a transport (no fallback to UDP or TCP). 
*  `tls_authentication`:  Use Strict Privacy i.e. require a TLS connection and authentication of the upstream
  * If Opportunistic mode is desired, simply remove the `tls_authentication: GETDNS_AUTHENTICATION_REQUIRED` field. In Opportunistic mode authentication of the nameserver is not required and fallback to clear text transports is permitted if they are in the `dns_transport_list`.
*  `tls_query_padding_blocksize`: Use the EDNS0 padding option to pad DNS queries to hide their size
*  `edns_client_subnet_private`: Use EDNS0 Client Subnet privacy so the client subnet is not sent to authoritative servers
* ` idle_timeout`:  Use an EDNS0 Keepalive idle timeout of 10s unless overridden by the server. This keeps idle TLS connections open to avoid the overhead of opening a new connection for every query.
*  `listen_address`: have the Stubby daemon listen on IPv4 and IPv6 on port 53 on the loopback address
*   `round_robin_upstreams`: Round robin queries across all the configured upstream servers. Without this option Stubby will use each upstream server sequentially until it becomes unavailable and then move on to use the next. 
*  `upstream_recursive_servers`: Use the NLnet labs test DNS Privacy Server for outgoing queries. In Strict Privacy mode, at least one of the following is required for each nameserver:
  *  `tls_auth_name`: This is the authentication domain name that will be verified against the presented certificate. 
  * `tls_pubkey_pinset`: The sha256 SPKI pinset for the server. This is also verified against the presented certificate. 

```
resolution_type: GETDNS_RESOLUTION_STUB
dns_transport_list: 
  - GETDNS_TRANSPORT_TLS
tls_authentication: GETDNS_AUTHENTICATION_REQUIRED
tls_query_padding_blocksize: 256
edns_client_subnet_private : 1
idle_timeout: 10000
listen_addresses:
  - 127.0.0.1
  -  0::1
round_robin_upstreams: 1
upstream_recursive_servers:
  - address_data: 185.49.141.38
    tls_auth_name: "getdnsapi.net"
    tls_pubkey_pinset:
      digest: "sha256"
       value: foxZRnIh9gZpWnl+zEiKa0EJ2rdCGroMWm02gaxSc9Q=
```

Additional privacy servers can be specified by adding more entries to the `upstream_recursive_servers` list above (note a separate entry must be made for the IPv4 and IPv6 addresses of a given server. More DNS Privacy test servers are listed [here](https://dnsprivacy.org/test_servers/).

A custom port can be specified by adding the `tls_port:` attribute to the `upstream_recursive_server` in the config file. 

More details can be found in the comments in the default configuration file and at https://dnsprivacy.org/wiki/display/DP/Configuring+Stubby

# Run Stubby


Simply invoke Stubby on the command line. By default it runs in the foreground, the `-g` flag runs it in the background. 

```sh
> sudo stubby
```

Or, to let it run as an unprivileged user:
```sh
> sudo setcap 'cap_net_bind_service=+ep' /usr/local/bin/stubby
> stubby
```

* Enable connection logging by using the `-l` flag. The logging is currently simplistic and simply writes to stdout. (We are working on making this better!)
* A custom configuration file can be specified using the -C flag.
* The pid file is /usr/local/var/run/stubby.pid

# Platform specific management
The Windows and macOS installers include scripts to run stubby as a managed daemon. We have basic support for using systemd to manage Stubby, see [systemd](https://github.com/getdnsapi/stubby/tree/master/systemd)

# Test Stubby

A quick test can be done by using dig (or your favourite DNS tool) on the loopback address

```sh
> dig @127.0.0.1 www.example.com
```

```sh
> getdns_query -s @127.0.0.1 www.example.com
```

# Modify your upstream resolvers

!!! <span class="glyphicon glyphicon-warning-sign"></span> Once this change is made your DNS queries will be re-directed to Stubby and sent over TLS! <br>
(You may need to restart some applications to have them pick up the network settings). <p>You can monitor the traffic using Wireshark watching on port 853.</p>

For Stubby to re-send outgoing DNS queries over TLS the system stub resolvers on your machine must be changed to send all the local queries to the loopback interface on which Stubby is listening. This depends on the operating system being run. It is useful to note your existing default nameservers before making this change!


## Linux/Unix systems

* Edit the /etc/resolv.conf file
* Comment out the existing *nameserver* entries
* Add the following (only add the IPv4 address if you don't have IPv6)
  ```sh
  nameserver 127.0.0.1
  nameserver ::1
  ```

## OS X

A script is provided with Stubby for easier configuration. From the command line you can do the following to switch all your queries to use Stubby

```sh
> sudo /usr/local/sbin/stubby-setdns-macos.sh
```

If you want to reset, just use:

```sh
> sudo /usr/local/sbin/stubby-setdns-macos.sh -r
```

which should pick up the default DHCP nameservers.

You can add /usr/local/sbin to your path to avoid having to type it above by doing
```
export PATH=/usr/local/sbin:$PATH
```

Or via the macOs System Settings GUI:

* Open *System Preferences &rarr; Network &rarr; Advanced &rarr; DNS*
* Use the '-' button to remove the existing nameservers
* Use the '+' button to add `127.0.0.1` and `::1` (only add the IPv4 address if you don't have IPv6)
* Hit 'OK' in the *DNS* pane and then 'Apply' on the *Network* pane

## Windows 8 and later

See the file windows/README.txt for a detailed description of building and using Stubby on windows.

Powershell scripts are provided in the the windows directory of the source code that can be used to update the system resolvers. 
Instructions for how to update the resolvers manually are provided are also provided - see https://dnsprivacy.org/wiki/display/DP/Windows+installer+for+Stubby 
Stubby has been reported to work on Windows 7, but we don't officially support it. 

## Notes:

* If Stubby works for a while but you then see failures from Stubby such as "None of the configured upstreams could be used to send queries on the specified transports" try restarting Stubby.
* If you are using a DNS Privacy server that does not support concurrent processing of TLS queries, you may experience some issues due to timeouts causing subsequent queries on the same connection to fail.

# Building getdns from source

Note that from getdns 1.1.3 stubby is included in the getdns code as a git submodule. Therefore stubby and getdns can be built together by following the
instructions in the getdns [README](https://github.com/getdnsapi/getdns/blob/develop/README.md) and setting the ``BUILD_STUBBY`` option.

## Logging/debugging when building from source

> **`--enable-debug-stub`**   If you do want to see _very_ detailed debug information as messages are processed then add the `--enable-debug-stub` option to the `configure` line above (not recommended for use with Stubby)


# Contributions

The contrib directory contains code kindly contributed by various people:

vapniks
Tom Matthews
CameronNemo


ChangeLog
*2203-07-26: vesrion 0.4.4-rc1
  * Fix issues with Windows build via PR £335 (thanks AstralStorm)
  * Remove `dnsovertls` servers from the man page
  * Fix broken wiki links (thanks Ayushka Partohap)

* 2022-12-22: version 0.4.3
  * Fix Issue #330 and PR#324: PrivateUsers=false needed in systemd
    stubby.service file for stubby to start.
    Thanks Archcan and Petr Menšík
  * PR #323: Reduce log messages when interface is offline.
    Thanks Russ Bubley and Andre Heider 

* 2022-08-19: version 0.4.2
  * Fix Issue #320: Stubby doesn't start without "log_level"
    entry. Thanks Johnny S. Lee

* 2022-08-19: Version 0.4.1
  * Several updates to the servers in the config file:
     * sinodun.dnsovertls*.com servers are removed and will be
       decommissioned in the near future. This leaves only the
       getdnsapi.net server as the default.
       A recommendation is made that users choose additional servers
       from the list available.
     * Additional Quad9 servers added (thanks pataquets).
     * LDN servers removed as the service is now stopped. 
     * Tidy up of remaining server data.
     * Change `comment` lines on Uncensored server data to more clearly
       be comments (not valid YAML)
     * Fix and improve descriptions of default values
     * getdnsapi.net DoT servers liston on port 443 now too
  * Fixes for parsing of quoted key values for `tls-ca-file`,
    `tls-curves-list` and `resolvconf`
  * Add Dockerfile into `contrib` - thanks to pataquets.
  * Strengthen version checking for LibIDN2 during the cmake step so
    LibIDN2 version 2.3.3 can be properly found (thanks jpbion).
  * Updated systemd/stubby.service file. Thanks Bruno Pagani
  * log level configurable in config file with log_level setting
    Command line setting (with -v or -l) overrules the config file
    setting

* 2021-06-04: Version 0.4.0
  * We announce the intention to remove the dnsovertls*.sinodun.com 
    servers from the default resolver list in the next release,
    see https://github.com/getdnsapi/stubby/issues/286 for details.
  * Add ability to run Stubby as a full Windows service
  * Update IPs for dot1.applied-privacy.net
  * Remove the secure.eu entries in the stubby config file
  * Bugfix: Make `run_in_foreground = 1` the default again
  * Support yaml config files that do not use quotes

* 2020-02-28: Version 0.3.0
  * Correct the applied-privacy.net details and remove ibksturm server in stubby.yml.example.
  * Include AppArmor profile (thanks CameronNemo)
  * Add `contrib` contributors to the README
  * Let scheduled task run indefinitely on Windows (thanks triatic)
  * Add `Restart=on-failure` to the systemd.service file (thanks appliedprivacy)

* 2019-12-20: Version 0.3.0-beta.1
  * Migration of build system to cmake.

* 2019-04-03: Version 0.2.6
 * Windows: use appropriate system and user configuration directories.
 * Windows: replace references to C:\Program Files with %PROGRAMFILES%.
 * Windows: use location of stubby.bat to find stubby.exe and stubby.yml.

* 2019-01-11: Version 0.2.5
 * Fix builds on Windows.
 * RFE getdnsapi/getdns#408: Document trust_anchors_backoff_time
   in stubby.yml.example. Thanks Jonathan Underwood
 * RFE #148: Document tls_ciphersuites, tls_cipher_list, tls_min_version
   and tls_max_version in stubby.yml.example. Thanks Jonathan Underwood
 * RFE #149: Added Google Public DNS to stubby.yml.example.
   Thanks Bruno Pagani

* 2018-12-21: Version 0.2.4
 * DNSSEC required with dnssec extension in example config
 * Removed the yeti servers from stubby.yml.example
 * Added the Foundation RESTENA servers in stubby.yml.example
 * Bugfix: only start Stubby when network is up
   Thanks Bruno Pagani

* 2018-05-11: Version 0.2.3
 * Bugfix #62 and #106: With systemd setups, make /run/stubby directory
   writeable for stubby user and include a "appdata_dir" directory
   in stubby.yml.example.
   Thanks Paul Wouters, eccgecko and Han Vinke
 * Update securedns.eu entries in stubby.yml.example
 * Added Cloudflare servers in stubby.yml.example
 * Added basic upstart script in contrib/upstart dir. Thanks vapniks
 * Bugfix #98: EDNS options that are handled internally should not
   be passed on through downstream.  Thanks Twisteroid Ambassador

* 2018-02-21: Version 0.2.2
 * Fixes and updates to the stubby.yml.config file. Add separate entries for
   servers that listen on port 443.
 * Additional logging of basic config on startup
 * -V option to show version
 * Added a man page

* 2017-12-18: Version 0.2.1
 * Fix use of logging on macos 10.11

* 2017-12-18: Version 0.2.0
 * Add Powershell scripts for Windows 7 that will update the IPv4 DNS resolvers.
 * Add Windows scripts to enable a Scheduled task for stubby
 * Add files to support a separate macOS GUI application to manage stubby
   https://dnsprivacy.org/wiki/x/CIBn
 * Add Quad9 details to the configuration file
 * Bugfix #48: Do not do native DNSSEC validation when cd bit was
   received (for example from an unbound forwarder), but DNSSEC
   validation was not enabled in the first place.
 * Bugfix getdnsapi/getdns#358: Parse config files given with the
   -C option that have an .yaml extension as YAML not JSON.
   Thanks Ollivier Robert

* 2017-11-03: Version 0.1.5
 * Add Windows installer package. Installer available at dnsprivacy.org
 * Fix to systemd file names (thanks ArchangeGabriel)
 * Add SPKI for Uncensored DNS (thanks woopstar)
 * Fix installation of stubby.yml file (thanks ArchangeGabriel)
 * Fix detection of platform for standalone build
 * Fix location of pid file installation
 * Update the stubby.yml file to contain details of all available servers. Only
   a small subset are enabled by default.

* 2017-10-20: Version 0.1.4
 *  '-i' option of stubby no longer tries to bind to the listen addresses so
    it can be run without requiring root privileges. Makes it easier to validate
    the configuration file syntax. 
 *  Fix incorrect IP addresses for some servers in the config file. Add note that

src/configfile.c
/*
 * Copyright (c) 2020, NLNet Labs, Sinodun
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 * * Redistributions of source code must retain the above copyright
 *   notice, this list of conditions and the following disclaimer.
 * * Redistributions in binary form must reproduce the above copyright
 *   notice, this list of conditions and the following disclaimer in the
 *   documentation and/or other materials provided with the distribution.
 * * Neither the names of the copyright holders nor the
 *   names of its contributors may be used to endorse or promote products
 *   derived from this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
 * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL Verisign, Inc. BE LIABLE FOR ANY
 * DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
 * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
 * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
 * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
 * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

#include "config.h"

#include <errno.h>
#if defined(STUBBY_ON_WINDOWS)
#include <winsock2.h>
#include <windows.h>
#include <shlobj.h>
#else
#include <pwd.h>
#include <unistd.h>
#endif
#include <string.h>

#include <getdns/getdns.h>
#include <getdns/getdns_extra.h>

#include "configfile.h"
#include "log.h"
#include "util.h"

#ifdef HAVE_GETDNS_YAML2DICT
getdns_return_t getdns_yaml2dict(const char *str, getdns_dict **dict);
#else
# include "yaml/convert_yaml_to_json.h"
# define getdns_yaml2dict stubby_yaml2dict
getdns_return_t
getdns_yaml2dict(const char *str, getdns_dict **dict)
{
        char *jsonstr;

        if (!str || !dict)
                return GETDNS_RETURN_INVALID_PARAMETER;

        jsonstr = yaml_string_to_json_string(str);
        if (jsonstr) {
                getdns_return_t res = getdns_str2dict(jsonstr, dict);
                free(jsonstr);
                return res;
        } else {
                return GETDNS_RETURN_GENERIC_ERROR;
        }
}
#endif

static char *make_config_file_path(const char *dir, const char *fname)
{
        size_t reslen = strlen(dir) + strlen(fname) + 1;
        char *res = malloc(reslen);

        if (res == NULL)
                return NULL;

        snprintf(res, reslen, "%s%s", dir, fname);
        return res;
}

#if defined(STUBBY_ON_WINDOWS)

static char *folder_config_file(int csidl)
{
        TCHAR szPath[MAX_PATH];

        if (!SUCCEEDED(SHGetFolderPath(NULL,
            csidl | CSIDL_FLAG_CREATE, NULL, 0, szPath)))
                return NULL;

        return make_config_file_path(szPath, "\\Stubby\\stubby.yml");
}

// %APPDATA%/Stubby/stubby.yml.
char *home_config_file(void)
{
        return folder_config_file(CSIDL_APPDATA);
}

char *system_config_file(void)
{
        return folder_config_file(CSIDL_PROGRAM_FILES);
}

char *system_service_config_file(void)
{
        int csidl = CSIDL_PROGRAM_FILES;
        TCHAR szPath[MAX_PATH];

        if (!SUCCEEDED(SHGetFolderPath(NULL,
            csidl | CSIDL_FLAG_CREATE, NULL, 0, szPath)))
                return NULL;

        return make_config_file_path(szPath, "\\Stubby\\stubbyservice.yml");        
}

#else

char *home_config_file(void)
{
        struct passwd *p = getpwuid(getuid());
        char *home = p ? p->pw_dir : getenv("HOME");
        if (!home)
                return NULL;
        return make_config_file_path(home, "/.stubby.yml");
}

char *system_config_file(void)
{
        return make_config_file_path(STUBBYCONFDIR, "/stubby.yml");
}

#endif

static const char *default_config =
"{ resolution_type: GETDNS_RESOLUTION_STUB"
", dns_transport_list: [ GETDNS_TRANSPORT_TLS"
"                      , GETDNS_TRANSPORT_UDP"
"                      , GETDNS_TRANSPORT_TCP ]"
", idle_timeout: 10000"
", listen_addresses: [ 127.0.0.1@53, 0::1@53 ]"
", tls_query_padding_blocksize: 256"
", edns_client_subnet_private : 1"
", round_robin_upstreams: 1"
"}";

static getdns_dict *listen_dict = NULL;
static getdns_list *listen_list = NULL;

static getdns_return_t parse_config(getdns_context *context,
		const char *config_str, int yaml_config, long *log_level)
{
        getdns_dict *config_dict;
        getdns_list *list;
        getdns_return_t r;
	uint32_t config_log_level;

        if (yaml_config) {
                r = getdns_yaml2dict(config_str, &config_dict);
                if (r == GETDNS_RETURN_NOT_IMPLEMENTED) {
                        /* If this fails then YAML is really not supported. Check this at
                           runtime because it could change under us..... */
                        r = getdns_yaml2dict(config_str, NULL);
                        if (r == GETDNS_RETURN_NOT_IMPLEMENTED) {
                                stubby_error("Support for YAML configuration files not available because\n"
                                             "the version of getdns used was not compiled with YAML support.");
                                return GETDNS_RETURN_NOT_IMPLEMENTED;
                        }
                }
        } else {
                r = getdns_str2dict(config_str, &config_dict);
        }
        if (r) {
                stubby_error("Could not parse config file %s, \"%s\"",
                    config_str, stubby_getdns_strerror(r));
                return r;
        }
        if (!(r = getdns_dict_get_list(
            config_dict, "listen_addresses", &list))) {
                if (listen_list && !listen_dict) {
                        getdns_list_destroy(listen_list);
                        listen_list = NULL;
                }
                /* Strange construction to copy the list.
                 * Needs to be done, because config dict
                 * will get destroyed.
                 */
                if (!listen_dict &&
                    !(listen_dict = getdns_dict_create())) {
                        stubby_error("Could not create listen_dict");
                        r = GETDNS_RETURN_MEMORY_ERROR;

                } else if ((r = getdns_dict_set_list(
                    listen_dict, "listen_list", list)))
                        stubby_error("Could not set listen_list");

                else if ((r = getdns_dict_get_list(
                    listen_dict, "listen_list", &listen_list)))
                        stubby_error("Could not get listen_list");

                (void) getdns_dict_remove_name(
                    config_dict, "listen_addresses");
        }
	if (!(r = getdns_dict_get_int(
	    config_dict, "log_level", &config_log_level))) {
		if (config_log_level > GETDNS_LOG_DEBUG)
			stubby_error("log level '%d' from config file is invalid or out of range (0-7)", (int)config_log_level);
		else
			*log_level = config_log_level;
		(void) getdns_dict_remove_name(
		    config_dict, "log_level");
	}
        if (!r && (r = getdns_context_config(context, config_dict))) {
                stubby_error("Could not configure context with "
                    "config dict: %s", stubby_getdns_strerror(r));
        }
        getdns_dict_destroy(config_dict);
        return r;
}

extern long log_level;
static getdns_return_t parse_config_file(getdns_context *context,
		const char *fn, long *config_log_level)
{
        FILE *fh;
        char *config_file = NULL;
        long config_file_sz;
        size_t read_sz;
        getdns_return_t r;

        if (!(fh = fopen(fn, "r")))
                return GETDNS_RETURN_IO_ERROR;

        if (fseek(fh, 0,SEEK_END) == -1) {
                perror("fseek");
                fclose(fh);
                return GETDNS_RETURN_IO_ERROR;
        }
        config_file_sz = ftell(fh);
        if (config_file_sz <= 0) {
                /* Empty config is no config */
                fclose(fh);
                return GETDNS_RETURN_IO_ERROR;
        }
        if (!(config_file = malloc(config_file_sz + 1))){
                fclose(fh);
                stubby_error("Could not allocate memory for \"%s\"", fn);
                return GETDNS_RETURN_MEMORY_ERROR;
        }
        rewind(fh);
        read_sz = fread(config_file, 1, config_file_sz + 1, fh);
        if (read_sz > (size_t)config_file_sz || ferror(fh) || !feof(fh)) {
                stubby_error("An error occurred while reading \"%s\": %s",
                             fn, strerror(errno));
                fclose(fh);
                free(config_file);
                return GETDNS_RETURN_IO_ERROR;
        }
        config_file[read_sz] = 0;
        fclose(fh);
        r = parse_config(context, config_file,
                         strstr(fn, ".yml") != NULL ||
                         strstr(fn, ".yaml") != NULL,
			 config_log_level);
        free(config_file);
        if (r == GETDNS_RETURN_GOOD
	&&  (          log_level <= GETDNS_LOG_DEBUG
	    || *config_log_level >= GETDNS_LOG_DEBUG))
                stubby_log(NULL,GETDNS_LOG_UPSTREAM_STATS, GETDNS_LOG_DEBUG,
                           "Read config from file %s", fn);
        return r;
}

void init_config(getdns_context *context, long *log_level)
{
        (void) parse_config(context, default_config, 0, log_level);
}

void delete_config(void)
{
        /* Listen list is a member of the dict, so will be destroyed */
        /* with the dict. */
        getdns_dict_destroy(listen_dict);
        listen_dict = NULL;
        listen_list = NULL;
}

int read_config(getdns_context *context, const char *custom_config_fn,
		int *validate_dnssec, long *log_level)
{
        char *conf_fn;
        int found_conf = 0;
        getdns_return_t r;
        getdns_dict *api_information = NULL;
        getdns_list *api_info_keys = NULL;
        int dnssec_validation = 0;

        (void) parse_config(context, default_config, 0, log_level);
        if (custom_config_fn) {
                if ((r = parse_config_file(context, custom_config_fn, log_level))) {
                        stubby_error("Could not parse config file "
                                     "\"%s\": %s", custom_config_fn,
                                     stubby_getdns_strerror(r));
                        return 0;
                }
        } else {
                conf_fn = home_config_file();
                if (!conf_fn) {
                        stubby_error("Error getting user config file");
                        return 0;
                }
                r = parse_config_file(context, conf_fn, log_level);
                if (r == GETDNS_RETURN_GOOD)
                        found_conf = 1;
                else if (r != GETDNS_RETURN_IO_ERROR)
                        stubby_error("Error parsing config file "
                                     "\"%s\": %s", conf_fn
                                     , stubby_getdns_strerror(r));
                free(conf_fn);
                if (!found_conf) {
                        conf_fn = system_config_file();
                        if (!conf_fn) {
                                stubby_error("Error getting system config file");
                                return 0;
                        }
                        r = parse_config_file(context, conf_fn, log_level);
                        if (r == GETDNS_RETURN_GOOD)
                                found_conf = 1;
                        else if (r != GETDNS_RETURN_IO_ERROR)
                                stubby_error("Error parsing config file "
                                             "\"%s\": %s", conf_fn
                                             , stubby_getdns_strerror(r));
                        free(conf_fn);
                }
                if (!found_conf)
                        stubby_warning("WARNING: No Stubby config file found... using minimal default config (Opportunistic Usage)");
        }

        if ((r = getdns_context_set_resolution_type(context, GETDNS_RESOLUTION_STUB))) {
                stubby_error("Error while trying to configure stubby for "
                             "stub resolution only: %s", stubby_getdns_strerror(r));
                return 0;
        }

        api_information = getdns_context_get_api_information(context);
        if (api_information
            && !dnssec_validation
            && !getdns_dict_get_names(api_information, &api_info_keys)) {
                size_t i;
                uint32_t value;
                getdns_dict *all_context;
                getdns_bindata *api_info_key;

                for ( i = 0
                    ; !dnssec_validation &&
                      !getdns_list_get_bindata(api_info_keys, i, &api_info_key)
                    ; i++) {
                        if ((  !strncmp((const char *)api_info_key->data, "dnssec_", 7)
                            || !strcmp ((const char *)api_info_key->data, "dnssec"))
                           && !getdns_dict_get_int(api_information, (const char *)api_info_key->data, &value)
                           && value == GETDNS_EXTENSION_TRUE)
                                dnssec_validation = 1;
                }
                getdns_list_destroy(api_info_keys);
                api_info_keys = NULL;
                if (   !dnssec_validation
                    && !getdns_dict_get_dict(api_information, "all_context"
                                                            , &all_context)
                    && !getdns_dict_get_names(all_context, &api_info_keys)) {
                        for ( i = 0
                            ; !dnssec_validation &&
                              !getdns_list_get_bindata(api_info_keys, i, &api_info_key)
                            ; i++) {
                                if ((  !strncmp((const char *)api_info_key->data, "dnssec_", 7)
                                    || !strcmp ((const char *)api_info_key->data, "dnssec"))
                                   && !getdns_dict_get_int(all_context, (const char *)api_info_key->data, &value)
                                   && value == GETDNS_EXTENSION_TRUE)
                                        dnssec_validation = 1;
                        }
                        getdns_list_destroy(api_info_keys);
                        api_info_keys = NULL;
                }
        }
        getdns_list_destroy(api_info_keys);
        getdns_dict_destroy(api_information);
        if ( validate_dnssec )
                *validate_dnssec = dnssec_validation;

        return 1;
}

char *config_get_api_info(getdns_context *context)
{
        getdns_dict *api_information = getdns_context_get_api_information(context);

        if (listen_dict && !getdns_dict_get_list(
                    listen_dict, "listen_list", &listen_list)) {

                (void) getdns_dict_set_list(api_information,
                                            "listen_addresses", listen_list);
        } else if (listen_list) {
                (void) getdns_dict_set_list(api_information,
                                            "listen_addresses", listen_list);

        } else if ((listen_list = getdns_list_create())) {
                (void) getdns_dict_set_list(api_information,
                                            "listen_addresses", listen_list);
                getdns_list_destroy(listen_list);
                listen_list = NULL;
        }

        return getdns_pretty_print_dict(api_information);
}

const getdns_list *get_config_listen_list(void)
{
        size_t list_count = 0;

        if ( listen_list )
                getdns_list_get_length(listen_list, &list_count);

        if ( list_count > 0 )
                return listen_list;
        else
                return NULL;
}
